### PR TITLE
Move towards a more composite approach for the options and outgoing contexts

### PR DESCRIPTION
**Pull request by andreasohlund**
Spiking a new approach to extend the contexts for the outgoing pipeline

### Highlights

* Less things are hardcoded into the "Unicastbus" and the XOptions
* Headers are now "set only" for the outgoing pipeline since there isn't really a use case to read the headers when sending. Any needed state should instead be stored in the "extension context"
* MessageId handling is now consolidated into one place and "read only" on the outgoing pipeline
* CorrelationId is now managed in a single place

### Composite api

This pull sets us of towards a more composite approach when it comes to adding features. This seems to lead to less coupling between "features".

#### Send/Reply/Publish options
The options is our way to expose ways to the user to control the above operations.
With this approach we no longer have to "hard code" those methods/properties on the option objects but rather use extension methods instead. I believe that this will allow use to adhear to SRP and OCP without sacrificing control. From a end user perspective nothing changes since everything will show up in their intellisense make it discoverable. 

From our own perspective one could argue that the code will be less readable since all options won't be in a single class. But I'd argue this is a price we can pay since you can still get this perspective by doing a "find usages" or the relevant option class

#### Outgoing context
Less things gets hardcoded and each feature can still choose to expose their data via extension methods. This pull show how the correlation id is exposed to the callback component

#### Incoming contexts
Not in scope here but I imaging that they will have the same benefits as the others

#### The soon the be handler contexts
Each feature should be able to expose relevant methods here as well. @danielmarbach I imaging that we would need to add some kind of context bag to make this happen?

#### BusConfiguration
Less things gets hardcoded on the object it self (see the outgoing headers). This means that there is less need for the `Configure` class which I believe can be removed soon. This also means that settings are stored in the settings dictionary and passed to the features that way instead of the mixed mode we have today, some in setting some hardcoded on `Configure`

#### External components

Since this model doesn't rely on things beeing "hardcoded" on various core classes external components, like the Callbacks, now have the same abilities to create api's as the core component has

#### New developers learning the codebase
This will definitely make it harder for new devs to learn the code base. I think we can mitigate this by doing a better job at documenting code design ideas like this one.

### Diagram
I've tried to draw the current model here:

https://drive.google.com/open?id=0Bxjcg5ArjglPV0hDM3h3STJEa1U&authuser=0

### Left to do
- [x] Move static headers into the "header" feature
- [x] Reach agreement in the team that this approach is a good idea
- [x] Create docs pull. This will include a new section on what correlation id is

### CODE REVIEW

**Comment by andreasohlund** [comment]
@SzymonPobiega ready for a first sneak peak

File: src/NServiceBus.Core/Extensibility/OptionExtensionContext.cs [hunk 1]
```diff
@@ -45,5 +45,24 @@ public class OptionExtensionContext
...
+        /// <returns></returns>
+        public T GetOrCreate<T>() where T:class,new()
+        {
+            T value;
+
+            if (TryGet(out value))
+            {
+                return value;
+            }
+            var newInstance = Activator.CreateInstance<T>();
```
Comment (SzymonPobiega) [inline src/NServiceBus.Core/Extensibility/OptionExtensionContext.cs]: why not `new T()`?

**Comment by SzymonPobiega** [comment]
What happened to `ConvertLegacyEnumResponseToLegacyControlMessageBehavior`?

**Comment by andreasohlund** [comment]
>What happened to ConvertLegacyEnumResponseToLegacyControlMessageBehavior?

Not needed anymore since we can now control if serialization should happen or not (so no need to remove headers, and set the body to byte[0]

File: src/NServiceBus.Core/Extensibility/OptionExtensionContext.cs (hunk 1 above)
Comment (andreasohlund) [inline src/NServiceBus.Core/Extensibility/OptionExtensionContext.cs]: Will fix

**Comment by andreasohlund** [comment]
@Particular/nservicebus I've fleshed out the descriptions of this one. Can you take a quick look and see if it makes sense and try to poke holes in it?

**Comment by danielmarbach** [comment]
Nitpicker: I find assertions in the handler a bad idead https://github.com/Particular/NServiceBus/pull/2735/files#diff-24cc6694d397b1869223645d282453d7R71

**Comment by SimonCropp** [comment]
yeah now that @danielmarbach mentions it that is prone to failure since if, due to some other bug, the handler is not executed we will get a false green

**Comment by danielmarbach** [comment]
Another idea: a lot of extensionpoint are relying on Dic<string,object> and we never questioned/discussed it. Should we brainstorm other approaches like Expando as well?

**Comment by andreasohlund** [comment]
>Another idea: a lot of extensionpoint are relying on Dic and we never questioned/discussed it. Should we brainstorm other approaches like Expando as well?

Good question, atm we have quite a few of those:

* Extendable options
* Pipeline "stash"
* SettingsHolder

and I imagine that we'll get a few more given the handler context and the send context. I'd say this should be handled in a different pull?

**Comment by andreasohlund** [comment]
>Nitpicker: I find assertions in the handler a bad idead

Yea I can see that. I was trying to minimize the code. I'll go through and change that to use the context and put the asserts in the test method

**Comment by andreasohlund** [comment]
I've created a rough drawing here https://drive.google.com/open?id=0Bxjcg5ArjglPV0hDM3h3STJEa1U&authuser=0  please chime in with suggestions for improvements

**Comment by SzymonPobiega** [comment]
I would add some info that `SetState...` happens during runtime while other happens during startup time

File: src/NServiceBus.Core/StaticHeaders/StaticHeadersConfigExtensions.cs [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+    /// </summary>
+    public static class StaticHeadersConfigExtensions
+    {
+        /// <summary>
+        /// Allows static outgoing headers to be set
+        /// </summary>
+        /// <param name="config">Configuration object</param>
+        /// <param name="key">The header key</param>
+        /// <param name="value">The header value</param>
+        public static void SetOutgoingHeader(this BusConfiguration config, string key, string value)
```
Comment (johnsimons) [inline src/NServiceBus.Core/StaticHeaders/StaticHeadersConfigExtensions.cs]: Maybe a better name ?
A lot of users get confused with the usage of these headers, should we name them something like `SetEndpointWideHeader` or something similar? Also api doco needs a bit more love IMO ;)

**Comment by SzymonPobiega** [comment]
I like the way it makes things more explicit.

File: src/NServiceBus.Core/StaticHeaders/StaticHeadersConfigExtensions.cs (hunk 2 above)
Comment (andreasohlund) [inline src/NServiceBus.Core/StaticHeaders/StaticHeadersConfigExtensions.cs]: +1 for a better name

`AddHeaderToAllOutgoingMessages(string key, string value)` ?

**Comment by andreasohlund** [comment]
@SzymonPobiega drawing updated to clearly separate runtime vs config time

**Comment by andreasohlund** [comment]
Docs pull created 

https://github.com/Particular/docs.particular.net/pull/639

will merge this one and complete the docs pull once the build is complete and uploaded to myget

**Pull request merged**